### ISSUE TITLE
[BUG] `make_fsdp_dtensor` views a rank-local buffer with the DTensor global trailing shape (rowwise-TP deadlock)

### ISSUE THREAD

**Issue by yuhezhang-ai** [NVIDIA/Megatron-LM#5786]
**Describe the bug**

Package: `megatron-fsdp` (PyPI). Affected version: 0.5.0 (latest published release at time of writing). Upstream source: NVIDIA/Megatron-LM, `megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py`. Symbol: `make_fsdp_dtensor` (module-level function). Location: `param_and_grad_buffer.py`, in the flat-buffer view block (near line 4726 in the installed 0.5.0 wheel).

There is a deterministic hang/deadlock at Megatron-FSDP initialization when combining Megatron-FSDP with Torch-native tensor parallelism (`DTensor`) that shards a parameter rowwise (i.e. a non-leading dimension differs per TP rank).

`make_fsdp_dtensor` builds the view shape for a rank-local flat buffer from the parameter's *global* trailing dimensions:

```python
if len(orig_param.shape) > 1:
    local_shape = (-1, *orig_param.shape[1:])
else:
    local_shape = (-1,)
```

For a `DTensor`, `orig_param.shape` is the global shape. Under rowwise TP the trailing dimensions differ per rank, so `orig_param.shape[1:]` is not the shape of the local shard held by this rank. The rank-local flat buffer only contains the rank-local shard, so the reshape/view against the global trailing shape is inconsistent across ranks: some DP/CP shards can perform the view while others cannot. The ranks that fail (or take a different branch) desynchronize from the uneven-DTensor metadata collectives that the remaining ranks enter, and the job deadlocks.

**Steps/Code to reproduce bug**

* Build a small model whose parameter is sharded rowwise across a TP group (e.g. a `Linear` sharded so a non-leading dimension differs per rank), wrapped as a Torch-native `DTensor`.
* Wrap it with Megatron-FSDP with `dp_size > 1` so the shard group is nontrivial.
* On any TP rank where the local trailing shape differs from the global trailing shape, `make_fsdp_dtensor` produces a view inconsistent with peer ranks, and the subsequent uneven-DTensor metadata collective deadlocks. With the `to_local().shape` fix, all ranks view their own shard and initialization completes.

**Expected behavior**

Initialization should complete on all ranks with each rank viewing its own shard. The fix is to derive the trailing dimensions from the rank-local tensor. For a `DTensor`, read the local shard via `to_local()`; a plain tensor's `.shape` is already rank-local:

```python
# helper
def _local_param_shape(param):
    tensor = param.to_local() if isinstance(param, DTensor) else param
    return tuple(tensor.shape)

local_param_shape = _local_param_shape(orig_param)
if len(local_param_shape) > 1:
    local_shape = (-1, *local_param_shape[1:])
else:
    local_shape = (-1,)
```

The same rank-local shape helper should be reused everywhere the flat buffer is viewed (see NVIDIA/Megatron-LM#5787, which repeats the identical global-shape mistake in the cached main-gradient path).

**Additional context**

Affected version: 0.5.0 (latest published release at time of writing). See NVIDIA/Megatron-LM#5787, which repeats the identical global-shape mistake in the cached main-gradient path.

**Comment by akoumpa** [comment]
Ping @cspades @shjwudp

**Comment by rehan243** [comment]
I hit a similar issue with DTensor and uneven sharding last week, turned out it was due to inconsistent views across ranks. Looking at the `make_fsdp_dtensor` function, I think the problem is with this line: `local_shape = (-1, *orig_param.shape[1:])`. For a rowwise TP'd DTensor, `orig_param.shape[1:]` is still the global shape, not the local shard shape. I've seen this before with PyTorch's DTensor, where you need to use `orig_param.to_local().shape` to get the actual local shard dimensions. 
To fix this, you could try changing it to something like:
```python
if isinstance(orig_param, DTensor):
    local_shape = (-1, *orig_param.to_local().shape[1:])
else:
    local_shape = (-1, *orig_param.shape[1:])
```
Are you using the latest PyTorch version with DTensor support?

**Comment by cspades** [comment]
@yuhezhang-ai @rehan243 I think I already understand this bug just reading it.

Since FSDP sharding comes after TP sharding, we need to make sure that the local shape going into the DTensor reflects the DP-TP sharded shape.

I suppose in this case the original parameter itself is already a DTensor, perhaps already sharded with TP (so the local shape is accurate) but the global shape is not TP-sharded, and we are passing a TP-unsharded DTensor global shape to the local shape of the DP-TP sharded DTensor.

Probably should walk through the logic of `make_fsdp_dtensor` to make sure it works for both MCore and OSS. I can look into this, but @yuhezhang-ai do me a favor, could you print out `orig_param` and `orig_param.shape` and `orig_param._local_tensor.shape` for me so I can make sure my understanding aligns with your bug? No reproduction details needed yet, just need to verify what a TP-sharded DTensor looks like.

@yuhezhang-ai @akoumpa Question, do we also need to support expert parallelism in DTensor? Does EP weight sharding use a single fused weight convention or do we have `expert{i}` weights?

**Comment by yuhezhang-ai** [comment]
Thanks @cspades — your understanding is exactly right. Here are the prints you asked for. Using a rowwise- vs colwise-TP linear (`[out=1024, in=1024]`, tp=2, torch 2.11/2.12 with DTensor):

```
[rowwise] placements=(Shard(dim=1),)  orig_param.shape (global)=(1024, 1024)  orig_param._local_tensor.shape (local)=(1024, 512)
[colwise] placements=(Shard(dim=0),)  orig_param.shape (global)=(1024, 1024)  orig_param._local_tensor.shape (local)=(512, 1024)
```

So for a **rowwise**-TP param, `orig_param.shape[1:]` is the TP-**global** trailing shape (`1024`), but the actual local shard trailing dim is `512`. `make_fsdp_dtensor`'s `local_shape = (-1, *orig_param.shape[1:])` therefore views the rank-local flat buffer with the global trailing size — the mismatch. Colwise happens to be fine because it shards dim 0, leaving `shape[1:]` equal to the local trailing dims; it's specifically the Shard-on-a-trailing-dim (rowwise) case that breaks. Reading `orig_param.to_local().shape` (as you and @rehan243 noted) is exactly what we do in our downstream patch and it resolves it.

For completeness, the combined DP×TP case from the actual failure looks like this (small dim-0 param, sharded on both meshes → empty tail shards):
```
DTensor(local_tensor=tensor([], size=(0, 1024)), device_mesh=DeviceMesh((dp=2, tp=2)), placements=(Shard(dim=0), Shard(dim=0)))
```

On your EP question: yes, we'd want DTensor EP support. Our experts use a **single fused weight** convention, not `expert{i}` weights — the grouped-expert projections are one `nn.Parameter` of shape `[n_experts, ...]` (e.g. `down_projs` = `[n_routed_experts, moe_inter_dim, expert_dim]`), run via grouped GEMM, and EP shards them on the expert dim (`Shard(0)`). So EP-in-DTensor would need to handle the fused 3-D expert tensor sharded on dim 0.

**Comment by akoumpa** [comment]
Hi @cspades , is anyone working on this? thanks

**Comment by cspades** [comment]
@akoumpa TBH I'm still very busy with inference and perf right now, will pass this to @wujingyue for whether to fix in V1 or support in V2.

**Comment by wujingyue** [comment]
Coincidentally, this issue came to my mind when I drafted https://github.com/NVIDIA/Megatron-LM/pull/6052 for MFSDP v2. So I'm pretty sure V2 will have the same bug. 

Let me see if there's a quick fix for V1...

**Comment by chengcuiping** [comment]
Hi @wujingyue — I reproduced both #5786 and #5787 on `main` at 9fa816bfd9dd994dc5acf6012fb4e841a56cfa4a, using the public `fully_shard` path with a torch-native rowwise-DTensor `Linear` on 4 GPUs (DP=2 × TP=2, FP32, `optim_grads_params`). The affected source file has the same blob on current `main`.

| Variant | Outcome |
| --- | --- |
| Unmodified base, rowwise `[8, 8]` | Hang — 2/2 runs timed out at 180 s |
| Only the `make_fsdp_dtensor` view fixed | #5787 reproduced — step 1 exact, step 2 fails |
| Both view sites fixed | Pass — 4 shapes × 5 steps, exact parity |

**Unmodified base, rowwise `[8, 8]`.** Two independent runs timed out after 180 s under `TORCH_DISTRIBUTED_DEBUG=INFO`, i.e. without `_ProcessGroupWrapper`. Both DP copies showed the same TP-axis split: TP rank 1 asserted on local shape `[2, 8]`, where `(Shard(0), Shard(1))` requires `[4, 4]`, while TP rank 0 remained blocked in the uneven-DTensor boundary synchronization. Colwise and replicated controls passed. Rowwise `[6, 10]` produced a deterministic view error instead.

**Only the `make_fsdp_dtensor` view changed to use the local trailing shape.** Initialization and optimizer step 1 matched the reference exactly; step 2 was confirmed to enter the cached `dist_main_grad` branch and failed at the repeated view in `update_main_grads`, reproducing #5787.

**Both view sites changed.** Four tested shapes completed five optimizer steps with exact parity against the TP-only reference (`max_abs = max_rel = 0`).

The tested scope was one synthetic `Linear` with DP=2 × TP=2; I did not test a full transformer, MoE, HSDP, or multi-parameter bucketing.

Are you already implementing the V1 quick fix you mentioned? If not, I'm happy to prepare one PR covering both issues, with a shared local-view-shape helper and a focused 4-rank, multi-step regression test. I would keep V2 and the separate pre-collective validation hardening out of scope. I also have the self-contained reproducer and per-rank logs ready to inline if useful.

**Comment by wujingyue** [comment]
@chengcuiping thanks. can you show me the repo for #5786? Try to minimize your repo -- fewer lines of code and fewer conditions and fewer parameters.

**Comment by chengcuiping** [comment]
Thanks @wujingyue  — here is the minimized reproducer:

https://github.com/chengcuiping/Megatron-LM/blob/repro/5786-minimal/tests/manual_tests/megatron_fsdp/repro_issue_5786.py

It contains one fixed case: a single `nn.Linear(8, 8, bias=False)` with one rowwise-DTensor parameter on a DP=2 × TP=2 mesh, passed to the public `fully_shard_model()` API. There is no optimizer, forward/backward pass, training loop, shape matrix, or CLI configuration.

```bash
PYTHONPATH=. CUDA_VISIBLE_DEVICES=0,1,2,3 timeout 60s \
  torchrun --standalone --nproc_per_node=4 \
  tests/manual_tests/megatron_fsdp/repro_issue_5786.py
```

On the unmodified `9fa816bfd9dd994dc5acf6012fb4e841a56cfa4a`, 2/2 runs produced 0/4 `INIT_OK`: ranks 1 and 3 asserted because the generated local shape was `[2, 8]` instead of `[4, 4]`, while ranks 0 and 2 remained blocked. `torchrun` then terminated the surviving ranks, so the visible result was exit 1 rather than a timeout.

Changing only the `make_fsdp_dtensor()` view to derive its trailing dimensions from the rank-local tensor made the same script complete with 4/4 `INIT_OK` in 2/2 runs. The branch contains only the reproducer and no production-code changes. I kept the optimizer-path problem from #5787 out of this #5786 reproducer.

**Comment by wujingyue** [comment]
Thanks -- the repro is really nice and focused. 

@yuhezhang-ai did you encounter this while using PyTorch’s [parallelize_module](https://github.com/pytorch/pytorch/tree/420de487c9de34177955f0ab0c13163f3eab6044/torch/distributed/tensor/parallel) API? More broadly, this seems to fall under TP composability. We support composition with Transformer Engine’s tensor parallelism, but have not yet prioritized support for PyTorch’s parallelize_module. I’m trying to understand how important that use case is for Automodel.

**Comment by yuhezhang-ai** [comment]
@wujingyue yes, via PyTorch's `parallelize_module` — that's the only TP mechanism in AutoModel's distributed path (DTensor `ColwiseParallel`/`RowwiseParallel` plans, including a rowwise vocab embedding, applied before the Megatron-FSDP wrap). We use TE for layer implementations, not for TP sharding.

That said, I wouldn't prioritize it on our account. I hit this during a trial, and we've since gone back to FSDP2 for the large runs — the few shipped Megatron-FSDP configs are all `tp_size=1`. Worth knowing too that #5786/#5787 alone wouldn't enable TP for us anyway: with both fixed on our fork, `tp>1` still fails at the first backward on the vocab-embedding grad placement (#5790).

**Comment by wujingyue** [comment]
Sounds good -- I'll try to support this in V2.
